Add `unist-util-ancestor` to list of utilities

gorango opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/syntax-tree/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/syntax-tree/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/syntax-tree/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Asyntax-tree&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

Adding a small utility to `readme` that helps with finding common ancestors of one or more unist nodes. It has come in handy a few times while working with ast trees so I thought I'd package it up. Here's hoping it might be helpful to someone else too!

P.S. I am open to renaming the package if desired.

<!--do not edit: pr-->

wooorm commented on readme.md:
```diff
@@ -431,6 +431,8 @@ unist:
 
 ### List of utilities
 
+*   [`unist-util-ancestor`](https://github.com/gorange/unist-util-ancestor)
```
```suggestion
*   [`unist-util-ancestor`](https://github.com/gorango/unist-util-ancestor)
```

wooorm commented:
Nice!

For performance reasons you might want to not use `unist-util-find` (in this way). And perhaps also `getParents` like this.
Right now, for every node, the whole tree is walked many many times.

If possible, it would be muuuch faster to walk the tree once/twice.
Here’s some pseudocode:

```js
const stacks = new Map()

visitParents(tree, (node, parents) => {
  if (nodesToFind.includes(node)) {
    stacks.set(node, parents)
  }
})

nodesToFind.forEach(node => {
  if (!stacks.has(node)) {
    throw new Error('unist-util-ancestor requires all nodes be contained in the tree')
  }
})

// Somethign like this?
let index = -1
let ancestor = tree

// This is broken, maybe you get the gist?
while (++index) {
  const nextAncestor = stacks.get(nodesToFind[0])[index]

  const shared = nodesToFind.every(node => stacks.get(node)[index] === nextAncestor)

  // To do: exit if there are no nodes left on stacks.

  if (shared) {
    ancestor = nextAncestor
  } else {
    break
  }
}
```

gorango commented:
Thank you so much for taking the time to review and provide this great feedback!

I have implemented your suggestions and removed the find util completely. Way cleaner :)

wooorm commented:
That looks great! Nice find as well on the things that I didn't get to. Glad it worked!
